src/http/authentication/authentication_types.h

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "enclave/rpc_context.h"
+#include "kv/tx.h"
+
+#include <memory>
+
+namespace ccf
+{
+  struct AuthnIdentity
+  {
+    virtual ~AuthnIdentity() = default;
+  };
+
+  class AuthnPolicy
+  {
+  public:
+    using OpenAPISecuritySchema = std::pair<std::string, nlohmann::json>;
+    static const OpenAPISecuritySchema unauthenticated_schema;
+
+    virtual ~AuthnPolicy() = default;
+
+    virtual std::unique_ptr<AuthnIdentity> authenticate(
+      kv::ReadOnlyTx& tx,
+      const std::shared_ptr<enclave::RpcContext>& ctx,
+      std::string& error_reason) = 0;
+
+    virtual void set_unauthenticated_error(
+      std::shared_ptr<enclave::RpcContext>& ctx, std::string&& error_reason)
+    {
+      ctx->set_error(
+        HTTP_STATUS_UNAUTHORIZED,
+        ccf::errors::AuthorizationFailed,
```

Comment:
Should probably be `InvalidAuthenticationInfo`.

Reply:
Done